File: kmp/shared/src/commonMain/kotlin/app/cadence/shared/domain/Occurrence.kt
package app.cadence.shared.domain

import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalTime
import kotlinx.datetime.daysUntil

private const val DAYS_PER_WEEK = 7

/**
 * Keyed by time as well as date: §03's `times[]` makes BPC-157 at 08:00 and 20:00 two
 * occurrences, and logging one doesn't log the other.
 */
data class Occurrence(
    val itemId: ProtocolItemId,
    val kind: ProtocolItemKind,
    val date: LocalDate,
    val time: LocalTime,
    val dose: Dose?,
    val status: OccurrenceStatus,
)

/**
 * Computed by comparing generated occurrences against logged events, never stored — §03's
 * L10 and «nothing derived is stored».
 */
enum class OccurrenceStatus { DONE, PENDING, MISSED, SCHEDULED }

/**
 * Week 1 is the seven days from the protocol's start; §03's second correction replaces the
 * prototype's hardcoded «week 4» against a frozen 31 May with this subtraction.
 */
fun cycleWeek(
    protocol: Protocol,
    date: LocalDate,
): Int? {
    val offset = protocol.startDate.daysUntil(date)
    if (offset < 0) return null
    val week = offset / DAYS_PER_WEEK + 1
    return if (week > protocol.weeks) null else week
}

/** One type, not three parameters that always travel together: they come from one fetch and are meaningless apart. */
data class ProtocolPlan(
    val protocol: Protocol,
    val items: List<ProtocolItem>,
    val phases: Map<ProtocolItemId, List<ProtocolPhase>>,
)

/**
 * §03: no materialized schedule table; generated on demand from protocol_items ×
 * protocol_phases. This is the whole schedule — Today and Schedule render the same call
 * (§03's seventh correction). `today` is a parameter, not read from a clock, so this can be
 * tested against a calendar.
 */
fun occurrencesFor(
    plan: ProtocolPlan,
    events: List<DoseEvent>,
    date: LocalDate,
    today: LocalDate,
): List<Occurrence> {
    // CANCELLED only, and the narrowing matters: a COMPLETED course is every patient after
    // twelve weeks, and blanking it would erase the dots from days they actually injected.
    if (plan.protocol.status == ProtocolStatus.CANCELLED) return emptyList()

    if (cycleWeek(plan.protocol, date) == null) return emptyList()

    return plan.items
        .filter { it.fallsOn(date) }
        .flatMap { item ->
            val dose = phaseDose(plan, item.id, date)
            item.times.map { time ->
                Occurrence(
                    itemId = item.id,
                    kind = item.kind,
                    date = date,
                    time = time,
                    dose = dose,
                    status = statusOf(item, date, time, events, today),
                )
            }
        }
}

/**
 * Null has three causes, one answer: cancelled, outside the window, or no phase covers
 * that week — offering a number in any case is the defect the step-4 review found on the
 * hero. One function, not two readings of `phases`: written twice, wizard and calendar
 * disagree the first time a band moves.
 */
fun phaseDose(
    plan: ProtocolPlan,
    itemId: ProtocolItemId,
    date: LocalDate,
): Dose? {
    if (plan.protocol.status == ProtocolStatus.CANCELLED) return null
    val week = cycleWeek(plan.protocol, date) ?: return null
    return plan.phases[itemId]?.firstOrNull { it.covers(week) }?.dose
}

/** Derived, not seeded: a stored copy goes stale the first time a protocol is edited. */
fun ProtocolItem.dosesPerWeek(): Double = scheduledDaysPerWeek() * times.size

/** Shared with [fallsOn], so the two readings of `cadence` cannot drift apart. */
private fun ProtocolItem.scheduledDaysPerWeek(): Double =
    when (cadence) {
        ProtocolCadence.DAILY -> DAYS_PER_WEEK.toDouble()
        ProtocolCadence.WEEKLY, ProtocolCadence.N_PER_WEEK -> daysOfWeek.size.toDouble()
    }

/** §03's `cadence` and `days_of_week[]`, read together. */
private fun ProtocolItem.fallsOn(date: LocalDate): Boolean =
    when (cadence) {
        ProtocolCadence.DAILY -> true
        ProtocolCadence.WEEKLY, ProtocolCadence.N_PER_WEEK -> date.dayOfWeek in daysOfWeek
    }

/**
 * Match must include item, date **and slot**: matching on item alone would mark every
 * remaining Sunday done after the first injection; without the slot, one BPC-157 event with
 * a null time closed both 08:00 and 20:00, telling a patient the evening dose was already done.
 */
private fun statusOf(
    item: ProtocolItem,
    date: LocalDate,
    time: LocalTime,
    events: List<DoseEvent>,
    today: LocalDate,
): OccurrenceStatus {
    val logged =
        events.any { event ->
            event.protocolItemId == item.id &&
                event.scheduledForDate == date &&
                event.scheduledForTime == time
        }
    if (logged) return OccurrenceStatus.DONE

    return when {
        date < today -> OccurrenceStatus.MISSED
        date == today -> OccurrenceStatus.PENDING
        else -> OccurrenceStatus.SCHEDULED
    }
}

File: api/internal/protocol/occurrence.go
package protocol

import (
	"slices"

	"github.com/SimonOsipov/cadence-app/api/internal/platform/civil"
)

const daysPerWeek = 7

// OccurrenceStatus is computed by comparing generated occurrences against logged events,
// never stored — §03's L10 and «nothing derived is stored».
type OccurrenceStatus string

const (
	StatusDone      OccurrenceStatus = "done"
	StatusPending   OccurrenceStatus = "pending"
	StatusMissed    OccurrenceStatus = "missed"
	StatusScheduled OccurrenceStatus = "scheduled"
)

// Occurrence is keyed by time as well as date: §03's `times[]` makes BPC-157 at 08:00 and
// 20:00 two occurrences, and logging one doesn't log the other.
type Occurrence struct {
	ItemID ProtocolItemID
	Kind   ItemKind
	Date   civil.Date
	Time   civil.Slot
	Dose   *Dose
	Status OccurrenceStatus
}

// CycleWeek counts week 1 as the seven days from the protocol's start; false means the date
// lies outside the course altogether.
func CycleWeek(p Protocol, d civil.Date) (int, bool) {
	offset := p.StartDate.DaysUntil(d)
	if offset < 0 {
		return 0, false
	}
	week := offset/daysPerWeek + 1
	if week > p.Weeks {
		return 0, false
	}
	return week, true
}

// OccurrencesFor is the whole schedule: §03 has no materialized schedule table, so Today and
// Schedule render the same call. `today` is a parameter rather than a clock reading, so this
// can be tested against a calendar.
func OccurrencesFor(plan Plan, logged []LoggedSlot, d, today civil.Date) []Occurrence {
	// CANCELLED only, and the narrowing matters: a COMPLETED course is every patient after
	// twelve weeks, and blanking it would erase the dots from days they actually injected.
	// The three read endpoints do not reach this branch today — ActivePlanFor selects the
	// active course, so a finished one is answered as no course at all. That is the step's
	// recorded divergence, not this function's rule.
	if plan.Protocol.Status == StatusCancelled {
		return nil
	}
	if _, ok := CycleWeek(plan.Protocol, d); !ok {
		return nil
	}

	var out []Occurrence
	for _, item := range plan.Items {
		if !fallsOn(item, d) {
			continue
		}
		dose := PhaseDose(plan, item.ID, d)
		for _, at := range item.Times {
			// A copy per slot: Kotlin's Dose is immutable and could be shared, a *Dose
			// cannot — one pointer across an item's slots ties BPC-157's morning and
			// evening occurrences to each other for anyone who writes through it.
			var own *Dose
			if dose != nil {
				value := *dose
				own = &value
			}
			out = append(out, Occurrence{
				ItemID: item.ID,
				Kind:   item.Kind,
				Date:   d,
				Time:   at,
				Dose:   own,
				Status: statusOf(item, d, at, logged, today),
			})
		}
	}
	return out
}

// PhaseDose returns nil for three causes with one answer: cancelled, outside the window, or
// no phase covering that week.
//
// It reads the phases unsorted while DoseBands and TitrationSteps sort them, so the two agree
// only while phases do not overlap — faithful to the Kotlin, and held by 000013's EXCLUDE
// rather than by anything here.
func PhaseDose(plan Plan, itemID ProtocolItemID, d civil.Date) *Dose {
	if plan.Protocol.Status == StatusCancelled {
		return nil
	}
	week, ok := CycleWeek(plan.Protocol, d)
	if !ok {
		return nil
	}
	for _, phase := range plan.Phases[itemID] {
		if phase.Covers(week) {
			dose := phase.Dose
			return &dose
		}
	}
	return nil
}

// CurrentDoseFor is the dose the course prescribes of a compound on a day, for callers that
// hold a drug rather than an item — the cabinet, whose vials name compounds.
//
// Nothing where two items name one compound, by the precedent OpenVialFor sets: 000013 has no
// unique index on the pair, so a course can carry an injection and a supplement of one drug,
// and answering either dose would put a number on a screen that half the prescription
// contradicts. An item that names the compound and carries no phase falls through to
// PhaseDose's own nil — Draft.Check requires phases of injections only, so a supplement
// naming a drug is a legal course.
func CurrentDoseFor(plan Plan, compound CompoundID, d civil.Date) *Dose {
	item := SoleItemFor(plan, compound)
	if item == nil {
		return nil
	}

	return PhaseDose(plan, item.ID, d)
}

// SoleItemFor is the one course position prescribing a compound, and nothing where two do.
//
// The ambiguity rule itself, so the readings that need the position rather than the dose —
// the cabinet's reorder hint divides by an item's own weekly rate — refuse on the same terms
// instead of carrying a second copy of it.
func SoleItemFor(plan Plan, compound CompoundID) *ProtocolItem {
	var sole *ProtocolItem
	for i, item := range plan.Items {
		if item.CompoundID == nil || *item.CompoundID != compound {
			continue
		}
		if sole != nil {
			return nil
		}
		sole = &plan.Items[i]
	}

	return sole
}

// DosesPerWeek is derived, not seeded: a stored copy goes stale the first time a protocol is
// edited.
func DosesPerWeek(item ProtocolItem) float64 {
	return scheduledDaysPerWeek(item) * float64(len(item.Times))
}

// Shared with fallsOn, so the two readings of `cadence` cannot drift apart.
func scheduledDaysPerWeek(item ProtocolItem) float64 {
	if item.Cadence == CadenceDaily {
		return daysPerWeek
	}
	return float64(len(item.DaysOfWeek))
}

func fallsOn(item ProtocolItem, d civil.Date) bool {
	if item.Cadence == CadenceDaily {
		return true
	}
	return slices.Contains(item.DaysOfWeek, d.Weekday())
}

// The match must include item, date and slot: matching on item alone would mark every
// remaining Sunday done after the first injection; without the slot, one BPC-157 event with
// a null time closed both 08:00 and 20:00.
func statusOf(item ProtocolItem, d civil.Date, at civil.Slot, logged []LoggedSlot, today civil.Date) OccurrenceStatus {
	for _, s := range logged {
		if s.ItemID == item.ID && s.Date == d && s.Time != nil && *s.Time == at {
			return StatusDone
		}
	}

	switch {
	case d.Before(today):
		return StatusMissed
	case d == today:
		return StatusPending
	default:
		return StatusScheduled
	}
}
